### PR TITLE
Fe customer bookings

### PR DESCRIPTION
**Pull request by wiroger9595**
### What this PR does (required):
- create customer booking UI page and mobile page 

### Screenshots / Videos (front-end only):

Screen Shot 2022-02-16 at 1 21 39 PM

Screen Shot 2022-02-16 at 1 22 01 PM

### Any information needed to test this feature (required):
- it need integrate with stripe API so it can proceed with the payment

### CODE REVIEW

File: client/src/components/TabPanel/TabPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+//   );
+// }
+
+// TabPanel.propTypes = {
+//   children: PropTypes.node,
+//   index: PropTypes.number.isRequired,
+//   value: PropTypes.number.isRequired,
+// };
+
+////////////////////////////////
```
Comment (ezeilo-su) [inline client/src/components/TabPanel/TabPanel.tsx:35]: Please let's remove unused code

File: client/src/components/TabPanel/TabPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    if (status === 'all') {
+      setMockData(CustomBookingMockData);
+      return;
+    }
+
+    const eachStatus = CustomBookingMockData.filter((item) => item.status === status);
+    setMockData(eachStatus);
+  };
+
+  /////////
```
Comment (ezeilo-su) [inline client/src/components/TabPanel/TabPanel.tsx:57]: This comment line is not needed here. Let's get rid of it :smiley:

File: client/src/components/TabPanel/TabPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    //   hidden={value !== index}
+    //   id={`simple-tabpanel-${index}`}
+    //   aria-labelledby={`simple-tab-${index}`}
+    // >
+    //   {value === index && (
+    //     <Box sx={{ p: 3 }}>
+    //       <Typography>{children}</Typography>
+    //     </Box>
+    //   )}
+    // </div>
```
Comment (ezeilo-su) [inline client/src/components/TabPanel/TabPanel.tsx:85]: Let's also get rid of this commented code

File: client/src/pages/Settings/CustomBooking/TicketList/TicketList.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  // });
+  // const [list, setList] = useState([]);
+
+  // useEffect(() => {
+  //   fetch('').then(async (res) => {
+  //     if (res.ok) {
+  //       setList(await res.json());
+  //     }
+  //   });
+  // }, [param]);
```
Comment (ezeilo-su) [inline client/src/pages/Settings/CustomBooking/TicketList/TicketList.tsx:54]: Let's remove!!!

**Review by ezeilo-su — CHANGES_REQUESTED** [review]
Good progress with this one. Please let's address those comments I left. Please also note that it's a requirement to always link the ticket being addressed in each PR. :smiley:

**Review by ezeilo-su — COMMENTED** [review]
Thank you for making the requested changes. Please do not merge until approved :smiley:

File: client/src/components/CustomBookingWraper/CustomBookingWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+interface CustomBookingWrapperProps {
+  customNavbar: JSX.Element | JSX.Element[];
+}
+
+const CustomBookingWrapper: React.FC<CustomBookingWrapperProps> = ({ customNavbar }) => {
+  return (
+    <Box
+      sx={{
+        background: '#fff',
```
Comment (moffatethan) [inline client/src/components/CustomBookingWraper/CustomBookingWrapper.tsx:12]: nit: use the colors library from MUI

File: client/src/components/TabPanel/TabPanel.tsx (hunk 2 above)
Comment (moffatethan) [reply client/src/components/TabPanel/TabPanel.tsx:30]: Did you forget this comment Rob?

File: client/src/components/TabPanel/TabPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        updateSnackBarMessage(data.error.message);
+      } else if (data.success) {
+        updateLoginContext(data.success);
+      } else {
+        // should not get here from backend but this catch is for an unknown issue
+        console.error({ data });
+        updateSnackBarMessage('An unexpected error occurred. Please try again');
+      }
+    });
+  };
```
Comment (moffatethan) [inline client/src/components/TabPanel/TabPanel.tsx:44]: I don't see this code being used anywhere in this file, remove?

File: client/src/context/useAuthContext.tsx [hunk 7]
```diff
@@ -8,7 +8,7 @@ import { User } from '../interface/User';
 import { Profile } from '../interface/Profile';
 
 interface IAuthContext {
-  profile: Profile | null | undefined;
+  profile: Profile | null | undefined | never | any;
```
Comment (moffatethan) [inline client/src/context/useAuthContext.tsx:11]: I prefer to not pollute this with `never` or `any`, look at your data coming back and make proper interfaces to capture different states of the profile so it excepts either of them

File: client/src/pages/Settings/CustomBooking/CustomBookingForm/CustomBooking.tsx [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+import TableHead from '@mui/material/TableHead';
+import TableContainer from '@mui/material/TableContainer';
+import Typography from '@mui/material/Typography';
+import Grid from '@mui/material/Grid';
+import ButtonBase from '@mui/material/ButtonBase';
+import Radio from '@mui/material/Radio';
+import RadioGroup from '@mui/material/RadioGroup';
+import FormControlLabel from '@mui/material/FormControlLabel';
+import FormControl from '@mui/material/FormControl';
+import FormLabel from '@mui/material/FormLabel';
```
Comment (moffatethan) [inline client/src/pages/Settings/CustomBooking/CustomBookingForm/CustomBooking.tsx:27]: Can we import on the individual exported components collectively. 

```
import { TableRow, TableHead, . . . } from '@mui/material';
```

Similar to what you have above, just combine then. It's messy when doing different types of imports.

File: client/src/pages/Settings/CustomBooking/CustomBookingForm/CustomBooking.tsx [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+    },
+  },
+});
+
+interface CustomBookingProps {
+  header: string;
+}
+
+const CustomBooking = ({ header }: CustomBookingProps): JSX.Element => {
+  const TAX_RATE = 0.0625;
```
Comment (moffatethan) [inline client/src/pages/Settings/CustomBooking/CustomBookingForm/CustomBooking.tsx:58]: Can we move this outside of the component

File: client/src/pages/Settings/CustomBooking/CustomBookingForm/CustomBooking.tsx [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+  function ccyFormat(num: number) {
+    return `${num.toFixed(2)}`;
+  }
```
Comment (moffatethan) [inline client/src/pages/Settings/CustomBooking/CustomBookingForm/CustomBooking.tsx:62]: This works for now but for integration we should use Stripe's last 4 digits they provide for the customers payment method

File: client/src/pages/Settings/CustomBooking/CustomBookingForm/CustomBooking.tsx [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+
+  function subtotal(items: readonly Row[]) {
+    return items.map(({ price }) => price).reduce((sum, i) => sum + i, 0);
+  }
+
+  const rows = [
+    createRow('Paperclips (Box)', 100, 1.15),
+    createRow('Paper (Case)', 10, 45.99),
+    createRow('Waste Basket', 2, 17.99),
+  ];
```
Comment (moffatethan) [inline client/src/pages/Settings/CustomBooking/CustomBookingForm/CustomBooking.tsx:88]: So, we're only ever going to have one row for the dog sitting service and nothing extra, let's just make one row here for the dog service

File: client/src/pages/Settings/CustomBooking/TicketList/TicketList.tsx [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+import { Box, CircularProgress, Grid, Link, Divider, Typography } from '@mui/material';
+import { NavLink, Redirect, Route, Switch, useHistory } from 'react-router-dom';
+import FormControlLabel from '@mui/material/FormControlLabel';
+import FormControl from '@mui/material/FormControl';
+import FormLabel from '@mui/material/FormLabel';
+import Chip from '@mui/material/Chip';
+import Radio from '@mui/material/Radio';
+import RadioGroup from '@mui/material/RadioGroup';
+import Tabs from '@mui/material/Tabs';
+import Tab from '@mui/material/Tab';
```
Comment (moffatethan) [inline client/src/pages/Settings/CustomBooking/TicketList/TicketList.tsx:11]: Same thing here as the previous comment 🙂

File: client/src/pages/Settings/Settings.tsx [hunk 13]
```diff
@@ -37,6 +38,11 @@ const settingsMenu = [
     to: '/profile/settings/billings',
     component: <StripeConnect />,
   },
+  {
+    name: 'Custom booking',
+    to: '/custom/booking',
+    component: <CustomBookingList header="Custom Booking" />,
+  },
```
Comment (moffatethan) [inline client/src/pages/Settings/Settings.tsx:45]: Not sure why this is on the settings page? This should be a page in the navbar for the customer.

File: client/src/pages/routes/route.ts [hunk 14]
```diff
@@ -57,11 +58,11 @@ const menuItems = [
...
   },
   {
-    item: 'Settings',
-    resource: '/profile/settings',
+    item: 'Custom Booking',
+    resource: '/custom/booking',
     canView: [AccountType.PET_SITTER, AccountType.PET_OWNER],
     authenticated: true,
-    component: Settings,
+    component: CustomBookingList,
```
Comment (moffatethan) [inline client/src/pages/routes/route.ts:65]: `Custom Booking` should be `Customer Booking`

File: client/src/pages/routes/route.ts [hunk 15]
```diff
@@ -57,11 +58,11 @@ const menuItems = [
     component: BookingPage,
   },
   {
-    item: 'Settings',
-    resource: '/profile/settings',
+    item: 'Custom Booking',
+    resource: '/custom/booking',
```
Comment (moffatethan) [inline client/src/pages/routes/route.ts:62]: `/customer/booking`

File: client/src/pages/routes/route.ts (hunk 14 above)
Comment (moffatethan) [reply client/src/pages/routes/route.ts:65]: Please update the component name too

File: client/src/pages/Settings/Settings.tsx (hunk 13 above)
Comment (moffatethan) [reply client/src/pages/Settings/Settings.tsx:45]: Make sure to move the components out of the settings folder to the pages

File: server/package.json [hunk 16]
```diff
@@ -27,6 +27,7 @@
     "morgan": "^1.10.0",
     "multer": "^1.4.4",
     "nodemon": "^2.0.15",
+    "parseur": "^0.1.0",
```
Comment (moffatethan) [inline server/package.json:30]: What's this for?

**Review by moffatethan — CHANGES_REQUESTED** [review]
Please bring back the sample.env - this is important for other developers on the project and potential future developers for onboarding. Shouldn't touch the file in this PR. 

Be mindful of import style too, would like to see group imports rather than individual imports when possible.

**Review by moffatethan — APPROVED** [review]

**Pull request merged**